Refactor Packer and PackerTool

lookeypl opened the pull request:
Resolves #1 - adds refactored Packer code (formerly known as Packer DLL) and new PackerTool.

This is a draft, first version of new Packer written almost entirely from ground up and still has missing features from #19 and #20. Additionally, Packer unit tests were written.

lookeypl committed:
Initial version of Packer console UI

This commit features initial version for Packer console UI. Initial
features consist of:
    * UI using callbacks as commands.
    * Simple exception system used in packerUI to inform user about
      unusual situations.
    * Preparation for config system for Packer to simplify use of some
      commands (not yet implemented).

libpacker API is still not implemented in command form. It will be
added on next, refactoring phase commit.

lookeypl committed:
Refactor existing Packer DLL code

Refactoring of existing Packer DLL. Refactoring major change list:
    * Project name change - Packer DLL -> Packer
    * API changes for more appropriate names
    * Return codes converted from macros to enum class PACK_RETURN
    * Most of Packer code is now multi-platform
    * Use existing functions with new Packer UI

lookeypl committed:
Corrections to PackerTool and Packer

* Some minor style issues (files ended with newline, etc.) were corrected
* Packer now keeps pointer to Reader/Writer inside unique_ptr
* Comments updated and corrected
* Unused code removed

Witek902 commented on nfEngine/Tools/PackerTool/packerTool.cpp:
```diff
@@ -0,0 +1,48 @@
+
+#include "packerToolCallbacks.hpp"
+
+using namespace std;
```
I suggest not doing that. Moreover, you are using "std::" in this file.

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/MD5.cpp:
```diff
@@ -0,0 +1,229 @@
+    mChunks[d] += 0x80 << mod*8;
+
+    // append size of message in bits to end of data
+    uint64 bitDataLength = (input.size()*8) % (0xffffffffffffffff);
```
What's "0xffffffffffffffff" for? uint64 can't be grater than that.

lookeypl replied:
Leftovers from legacy code (I don't remember why I thought this was a good idea). Removed.

Witek902 commented on nfEngine/nfCommon/MD5.cpp:
```diff
@@ -0,0 +1,229 @@
+    MD5DataSet data(input);
+
+    // do complete round of MD5 calculation for each set of 16 chunks
+    for (unsigned int i=0; i < data.mChunkSize; i+=16)
```
Style (spaces between opearators). Scan modified files with AStyle

lookeypl replied:
Done. Accidentially AStyle fixed ThreadPool-related files as well.

Witek902 commented on nfEngine/nfCommon/MD5.cpp:
```diff
@@ -0,0 +1,229 @@
+    }
+
+    // TODO: in memory data is aligned byte-to-byte correctly. However when kept in uint32 we have
+    //       Big/Little Endian difference. Question - is this shifting necessary for us?
```
This swapping can be done only for display puropses. I suggest writing separate macro/function like SwapBytes in separate file (nfCommon/Bit.hpp or something like that).

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/MD5.cpp:
```diff
@@ -0,0 +1,229 @@
+
+void MD5Hash::Set(uint32 h0, uint32 h1, uint32 h2, uint32 h3)
+{
+    this->h[0] = h0;
```
"this" not needed.

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/MD5.h:
```diff
@@ -0,0 +1,88 @@
+#include "nfCommon.h"
+
+// basic MD5 bitwise operations
+#define MD5_F(x, y, z) (((x) & (y)) | ((~x) & (z)))
```
These should be moved to cpp file to reduce namespace pollution.

Witek902 commented on nfEngine/nfCommon/MD5.h:
```diff
@@ -0,0 +1,88 @@
+    void Print() const;
+
+private:
+    uint32 h[4];
```
Consider switching to SSE's __m128i

lookeypl replied:
OK, comment with TODO added

Witek902 commented on nfEngine/nfCommon/Packer/packer.cpp:
```diff
@@ -0,0 +1,51 @@
+/**
```
Write boilerplates in doxygen style. Use "@" everywhere. Add missing boilerplates in other files.

Witek902 replied:
Packer.cpp

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/Packer/packer.hpp:
```diff
@@ -0,0 +1,29 @@
+namespace NFE {
+namespace Common {
+
+NFCOMMON_API PACK_RESULT Packer_CreateWriter(PackWriter** retwriter);
```
Are these really needed? User (packTool and engine) can't just do "new PackReader"?

lookeypl replied:
Removed

Witek902 commented on nfEngine/nfCommon/Packer/packerDefines.hpp:
```diff
@@ -0,0 +1,60 @@
+namespace Common {
+
+// return codes used by Packer
+enum class PACK_RESULT : uint32
```
Write enum names in CamelStyle

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/Packer/packerDefines.hpp:
```diff
@@ -0,0 +1,60 @@
+enum class PACK_RESULT : uint32
+{
+    OK = 0,
+    UNINITIALIZED,
```
Enum values also

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/Packer/packerElement.hpp:
```diff
@@ -0,0 +1,28 @@
+{
+    MD5Hash mHash;
+    std::string FilePath;
+    size_t FilePos;
```
What about support for files >4GB on 32-bit version?

lookeypl replied:
FILE does not support them - the limit is size_t accuracy. This will be added when #21 will be done.

Witek902 replied:
OK

Witek902 commented on nfEngine/nfCommon/Packer/packerReader.hpp:
```diff
@@ -0,0 +1,41 @@
+    NFCOMMON_API size_t GetFileCount() const;
+    NFCOMMON_API uint32 GetFileVersion() const;
+
+#ifdef NFCOMMON_EXPORTS
```
There is no point in hiding private section. This makes usage of "new" operator impossible.

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/Packer/packerWriter.cpp:
```diff
@@ -0,0 +1,182 @@
+    return PACK_RESULT::OK;
+}
+
+PACK_RESULT PackWriter::AddFile(const std::string& FilePath, const std::string& VFSFilePath)
```
Write function for adding Buffer object as well. This will be useful for packing compiled shaders.

Witek902 commented on nfEngine/nfCommon/Packer/packerWriter.cpp:
```diff
@@ -0,0 +1,182 @@
+        do
+        {
+            readCount = fread(buffer, sizeof(unsigned char), PACKER_DEF_BUFFER_SIZE, p_AppendedFile.get());
+            fseek(pFile.get(), 0, SEEK_END);
```
Not needed - fwrite will increment cursor.

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/Packer/packerWriter.hpp:
```diff
@@ -0,0 +1,42 @@
+    NFCOMMON_API size_t GetFileCount() const;
+    NFCOMMON_API const std::string& GetPAKName() const;
+
+#ifdef NFCOMMON_EXPORTS
```
The same as in PackReader.

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommonTest/main.cpp:
```diff
@@ -13,6 +13,7 @@ void SetUpCurrentDirectory()
 
 int main(int argc, char* argv[])
 {
+    //PROFILE_COMMAND_STATUS profstat = StopProfile(PROFILE_GLOBALLEVEL, PROFILE_CURRENTID);
```
Remove

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/Packer/packerWriter.cpp:
```diff
@@ -0,0 +1,182 @@
+namespace Common {
+
+PackWriter::PackWriter() {}
+PackWriter::PackWriter(const std::string& ArchiveName)
```
Begin function arguments with lowercase.

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/Packer/packerWriter.hpp:
```diff
@@ -0,0 +1,42 @@
+namespace NFE {
+namespace Common {
+
+class PackWriter
```
Make whole class NFCOMMON_API

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/MD5.cpp:
```diff
@@ -0,0 +1,229 @@
+
+MD5DataSet::~MD5DataSet()
+{
+    if (mChunks != nullptr)
```
Use smart pointer

lookeypl replied:
Since mChunks is an array, I'll rather use vector instead of unique_ptr

Witek902 commented on nfEngine/Tools/PackerTool/packerToolCallbacks.cpp:
```diff
@@ -0,0 +1,327 @@
+//HELP MESSAGES//
+/////////////////
+const std::string helpStr_load = "\
+Loads PAK archive into memory.\n\
```
C++11 "raw string literal" can be used here to improve readability.

lookeypl replied:
Done

Witek902 commented on nfEngine/Tools/PackerTool/packerToolUI.cpp:
```diff
@@ -0,0 +1,206 @@
+    //Throw("Reading config failed.");
+}
+
+void packerToolUI::AddCommand(std::string commandString, CommandFunc command)
```
const std::string&

lookeypl replied:
Done

Witek902 commented on nfEngine/Tools/PackerTool/packerToolCommand.hpp:
```diff
@@ -0,0 +1,34 @@
+#include <vector>
+#include <string>
+
+typedef void (*CommandFunc)(std::string& cmdString);
```
Use std::function

lookeypl replied:
Done

Witek902 commented on nfEngine/Tools/PackerTool/packerToolCommand.hpp:
```diff
@@ -0,0 +1,34 @@
+    void SetCommandHelpMessage(std::string helpMessage);
+
+private:
+    CommandFunc m_command;
```
mCommand

lookeypl replied:
Done, for other variables as well

Witek902 commented on nfEngine/Tools/PackerTool/packerToolCommand.hpp:
```diff
@@ -0,0 +1,34 @@
+private:
+    CommandFunc m_command;
+
+    std::string m_commandString;
```
mCommandString, check other places

lookeypl replied:
Done

Witek902 commented on nfEngine/Tools/PackerTool/packerToolDefines.cpp:
```diff
@@ -0,0 +1,31 @@
+const std::string s_endColoring = "\033[0m";
+#endif
+
+void PrintColored(std::string text, ConsoleColor foreground)
```
I suggest moving that to separate unit in nfCommon. It may be useful in other tools/logger/etc.

lookeypl replied:
Done

Witek902 commented on nfEngine/Tools/PackerTool/packerToolDefines.hpp:
```diff
@@ -0,0 +1,67 @@
+#pragma once
+
+#ifdef WIN32
```
Move all external includes to precompiled header

lookeypl replied:
Done

Witek902 commented on nfEngine/Tools/PackerTool/packerToolDefines.hpp:
```diff
@@ -0,0 +1,67 @@
+//Exception definition
+#include "packerToolException.hpp"
+
+#define SAFE_MEMFREE(x) if(x) { delete x; x = 0; }
```
do ... while (0)

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/MD5.cpp:
```diff
@@ -0,0 +1,229 @@
+namespace NFE {
+namespace Common {
+
+MD5DataSet::MD5DataSet(const std::string& input):
```
Initializer list style

lookeypl replied:
Not needed anymore, since mChunkSize disappeared and mChunks became a vector.

Witek902 commented on nfEngine/nfCommon/Packer/packerElement.hpp:
```diff
@@ -0,0 +1,28 @@
+namespace NFE {
+namespace Common {
+
+struct packElement
```
PackElement

lookeypl replied:
I'll prefer PackerElement, since all the classes are converted into Packer*. Done

Witek902 commented on nfEngine/nfCommonTest/PackerBasicTest.cpp:
```diff
@@ -0,0 +1,122 @@
+        }
+    }
+
+    PackReader* reader;
```
mReader

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommonTest/PackerComplexTest.cpp:
```diff
@@ -0,0 +1,332 @@
+
+    // create buffer and read data
+    correctData->Create(correctDataSize);
+    fread(correctData->GetData(), sizeof(unsigned char), correctDataSize, pFile);
```
Check return value

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommonTest/PackerComplexTest.cpp:
```diff
@@ -0,0 +1,332 @@
+
+        // create buffer and read data
+        correctData->Create(correctDataSize);
+        fread(correctData->GetData(), sizeof(unsigned char), correctDataSize, pFile);
```
Check return value

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommonTest/PackerComplexTest.cpp:
```diff
@@ -0,0 +1,332 @@
+
+        // create buffer and read data
+        correctData->Create(correctDataSize);
+        fread(correctData->GetData(), sizeof(unsigned char), correctDataSize, pFile);
```
Check return value

Witek902 commented on nfEngine/nfCommon/Packer/packer.hpp:
```diff
@@ -0,0 +1,29 @@
+/**
```
Packer.hpp

lookeypl replied:
Done

Witek902 commented on nfEngine/nfCommon/Packer/packerDefines.hpp:
```diff
@@ -0,0 +1,60 @@
+/**
```
PackerDefines.hpp

lookeypl replied:
Done

Witek902 commented:
Review done.

lookeypl committed:
Merge branch 'devel' into packer1

Conflicts:
	nfEngine/Tools/Packer/packer.hpp
	nfEngine/Tools/PackerTool/packerTool.vcxproj
	nfEngine/nfCommon/nfCommon.vcxproj
	nfEngine/nfCommon/nfCommon.vcxproj.filters
	nfEngine/nfCommonTest/nfCommonTest.vcxproj
	nfEngine/nfCommonTest/nfCommonTest.vcxproj.filters

lookeypl committed:
[CRFIX 1/3] Additional after-merge corrections

This is just a set of after-merge corrections which were left out due to
overwhelming amount of conflicts and differences between branches. Correct
CRFIX after review will come up next.

lookeypl committed:
[CRFIX 2/3] Code review fixes

This commit covers most of Code Review fixes - basic changes like fixing
style mistakes, missing returns, etc.

lookeypl committed:
[CRFIX 3/3] Add stub function for adding Buffer to PAK

This commit adds a stub AddFile overload to PackerWriter which takes
Buffer object as argument. Implementing multiple resource types in Packer
requires some reorganization and will be done in separate commit.

lookeypl committed:
[CRFIX]
